initialization/initialization.go

```diff
@@ -170,17 +213,68 @@ func (init *Initializer) Initialize(ctx context.Context) error {
 
 	numLabels := uint64(init.opts.NumUnits) * uint64(init.cfg.LabelsPerUnit)
 	fileNumLabels := numLabels / uint64(init.opts.NumFiles)
+	difficulty := shared.PowDifficulty(numLabels)
+	batchSize := uint64(config.DefaultComputeBatchSize)
 
 	init.logger.Info("initialization: starting to write %v file(s); number of units: %v, number of labels per unit: %v, number of bits per label: %v, datadir: %v",
 		init.opts.NumFiles, init.opts.NumUnits, init.cfg.LabelsPerUnit, init.cfg.BitsPerLabel, init.opts.DataDir)
 
 	for i := 0; i < int(init.opts.NumFiles); i++ {
-		if err := init.initFile(ctx, uint(init.opts.ComputeProviderID), i, numLabels, fileNumLabels); err != nil {
+		if err := init.initFile(ctx, i, batchSize, numLabels, fileNumLabels, difficulty); err != nil {
 			return err
 		}
 	}
 
-	return nil
+	if init.nonce != nil {
+		return nil
+	}
+
+	init.logger.Info("initialization: no nonce found while computing leaves, continue searching")
+	if init.lastPosition == nil || *init.lastPosition < numLabels {
+		init.lastPosition = new(uint64)
+		*init.lastPosition = numLabels
+	}
+
+	// continue searching for a nonce
+	// TODO(mafa): PM-195 depending on the difficulty function this can take a VERY long time, with the current difficulty function
+	// ~ 37% of all smeshers won't find a nonce while computing leaves
+	// ~  2% of all smeshers won't find a nonce even after checking 4x numLabels
+	for i := *init.lastPosition; i < math.MaxUint64; i += batchSize {
+		defer init.saveMetadata()
```

Comment:
This will be called (in a reversed order) at the end of the function, not the loop clause. 

How about doing `go init.saveMetadata()` instead? It's best-effort only anyway.

Reply:
I moved `defer init.saveMetadata()` outside of the for loop. `go init.saveMetadata()` would be racy here -> `init.lastPosition` and `init.nonce` might update between calling `go init.saveMetadata()` and it executing.

The reason I'm using `defer` is because I want to update that the metadata file no matter how the loop is ended; by context cancellation, when an error occurs or when the nonce is found.

Reply:
Looks fine now, together with https://github.com/spacemeshos/post/pull/67#discussion_r1034439363.